### PR TITLE
AzureMonitor: Alerting for Azure Application Insights

### LINKED ISSUE #15153: Convert Azure Application Insights datasource to Go
The Azure Monitor datasource plugin integrates against 4 Azure API's.

To support alerting for the Azure Application Insights service, the [datasource](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/grafana-azure-monitor-datasource/app_insights/app_insights_datasource.ts) class for the Azure Application Insights API needs to be converted from TypeScript to Go.

It should be structured in a similar way to Stackdriver: https://github.com/grafana/grafana/blob/master/pkg/tsdb/stackdriver/stackdriver.go

### PR DESCRIPTION
**Pull request by ChadNedzlek**
Allows for alerting of Application Insights data source

Closes: #15153

This also changes the names of the App Insights datasource property name to
match other datasources to make code sharing easier, it more closely matched the
existing AzureMonitor names in the same plugin, as well as other raw-query type
datasources.

### CODE REVIEW

**Comment by ChadNedzlek** [comment]
I do wonder if changing things like "xaxis" to "timeColumn" to look more like the other query style datasources is something I should avoid.  I'll happily make that change if someone thinks that's better.

I'm also not clear why the checks failed... the error message doesn't actually say what's wrong.

**Comment by ChadNedzlek** [comment]
@torkelo or @sunker I'd appreciate some eyes on this, especially if I'm off base with the property renames to make the backend code more logical.

**Comment by ChadNedzlek** [comment]
@daniellee Would you be the right person to look at this?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts]: Typo here

**Comment by sunker** [comment]
Just found a minor typo. Also it would be great if you could update the [readme](https://github.com/grafana/grafana/blob/master/docs/sources/features/datasources/cloudwatch.md) and point out that App Insights also supports alerting. 

I think this looks great, but I would like to get @daniellees opinion before merging since he's the one who's been working with this datasource.

**Review by sunker — CHANGES_REQUESTED** [review]

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts
Comment (ChadNedzlek) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts]: Fixed and added a test to make sure the migration is doing what I expected

**Comment by ChadNedzlek** [comment]
@sunker Updated the readme to indicate alerting is working.  I copied the blob from the Azure Monitor part.

File: public/app/features/panel/query_editor_row.ts
Comment (daniellee) [inline public/app/features/panel/query_editor_row.ts]: Can you explain what this is for? Is this a fix for the toggle to edit mode when the query mode is not Application Insights?

**Review by daniellee — CHANGES_REQUESTED** [review]
Looks really good. The change in the partial has to be fixed though before merging.

File: pkg/tsdb/azuremonitor/applicationinsights-datasource.go
Comment (daniellee) [inline pkg/tsdb/azuremonitor/applicationinsights-datasource.go]: It would be better if the response from Application Insights metrics could be typed instead. This could be difficult as the format of the response changes a lot. We could also merge first and fix afterwards.

File: pkg/tsdb/azuremonitor/applicationinsights-datasource.go
Comment (ChadNedzlek) [reply pkg/tsdb/azuremonitor/applicationinsights-datasource.go]: Fix this one (and a few others in the file)

File: pkg/tsdb/azuremonitor/applicationinsights-datasource.go
Comment (ChadNedzlek) [reply pkg/tsdb/azuremonitor/applicationinsights-datasource.go]: I initially tried that, but there are two things that I couldn't resolve.  Firstly, the answer varies _wildly_ depending in if you are asking for A) intervaled values or just a single datapoint and B) if you've added a dimension or not. Secondly, the _keys_ of the JSON values are intermixed. The result looks like
```json
{
  "start": "2019...",
  "end": "2019...",
  "NAME-OF-METRIC": ...
```

And I couldn't figure out a good typed way to handle "extra keys" that didn't just drop them.

I'm new to Go, so it's possible I missed something, and I'd be happy to change it if those problems are surmountable.

File: public/app/features/panel/query_editor_row.ts
Comment (ChadNedzlek) [reply public/app/features/panel/query_editor_row.ts]: Sorry, I was trying to fix the fact that the little penil icon doesn't work until you switch away and back. But I don't think it quite worked (there is something else caching the wrong value higher up in the chain I couldn't find). I'll revert this.

There is a bug that when you first switch to "Application Insights" in he dropdown, you can't go into raw query mode, because the fact that "Azure Monitor" doesn't have a text edit mode is cached already, and it was _really_ getting in the way of my validation. I initially thought I'd introduced it, but I was incorrect.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/partials/query.editor.html
Comment (ChadNedzlek) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/partials/query.editor.html]: :-( That's my mistake. I got hit with a huge merge conflict in this file that got reverted... it took me a long time to recover from that, and apparently I still missed a few. Since this file has no compile time type safety, I appreciate your detailed review of it.

**Comment by ChadNedzlek** [comment]
@daniellee Thank you for the review. I think I've addressed all but one of your comments.  If there is some json serialization method available I can't figure out, I'd love to change it (having to dip into those maps over and over and deal with the type coercion was extremely finicky and painful, so I'd hate to inflict that on future maintainers if it could be avoided)

File: pkg/tsdb/azuremonitor/applicationinsights-datasource.go
Comment (daniellee) [reply pkg/tsdb/azuremonitor/applicationinsights-datasource.go]: Yes, I guessed that was the reason you went with SimpleJson. I can have a look and see if it is possible to do it but I can do that after merging. We did manage to do it for other datasources but none of them was as fickle as the App Insights data format.

**Review by daniellee — APPROVED** [review]
LGTM

Will follow up with an issue for the typing and an issue to refactor the UI for Application Insights Logs/Analytics.

Thanks @ChadNedzlek for all your work!

**Review by sunker — APPROVED** [review]

**Comment by jperlope** [comment]
@ChadNedzlek 
I've tested the new version 6.4.2 (2019-10-08) and I've realized that the error message "tsdb.HandleRequest() error Alerting not supported for Application Insights" is still happen.
Untitled

Is this the expected behaviour? Or this feature is still developing? 

Thanks @ChadNedzlek for all your work!

**Comment by marefr** [comment]
@jperlope It will be included in Grafana v6.5. You can test a nightly build until that.

**Pull request merged**
